### ISSUE TITLE
deployment is painfully slow

### ISSUE THREAD

**Issue by loveshack** [gc3-uzh-ch/elasticluster#611]
Deploying on AWS is painfully slow for me.  It takes ~30 minutes for two nodes on c5.large instances with current development source and an AWS HPC AMI, i.e. with various relevant packages already installed.

Sitting watching, it seemed most of that is in package updates and installation, which just shouldn't take that long (and doesn't, comparing with a simple vagrant setup, for instance).  I discovered that's because of loops over packages. I modified at least the most important instances of loops to use a package list instead.  That reduced the time to ~10 minutes.  I could supply changes, but I'm not an ansible expert, so probably haven't done it appropriately.

**Comment by riccardomurri** [comment]
If you are installing from source, can you please install Python package [`ara`][1] (`pip install ara`) into the same virtualenv where ElastiCluster is running?  That will create a SQLite DB with data about every Ansible task executed, which I can then use for profiling.

Don't hold great expectations though :-( Ansible is quite slow especially in copying files and installing packages so we won't be able to shave much.  On the plus side, though, setup time is the same for any number of nodes up to `ANSIBLE_FORK`, so you get ~30 minutes for 2 through 2*num_cpus nodes...

[1]: https://pypi.org/project/ara/

**Comment by loveshack** [comment]
> If you are installing from source, can you please install Python
> package [`ara`][1] (`pip install ara`) into the same virtualenv where
> ElastiCluster is running?  That will create a SQLite DB with data
> about every Ansible task executed, which I can then use for profiling.

I can do, but is there much point without fixing the factor of three
from package update/installation that's obvious by eye?  Also I don't
know how much variation there might be in the timing, though it seemed
reasonably consistent in two or three goes.

> Don't hold great expectations though :-( Ansible is quite slow
> especially in copying files and installing packages so we won't be
> able to shave much.

I noticed it's slow, but the main slowness in package installation seems
to be how it's done.  (Even keeping loops, there's redundancy from
mentioning some redundantly in different places and others that are in
the base distribution; also some seem odd choices for a default
installation.)

**Comment by riccardomurri** [comment]
Commits b704938..d6f312e8 reap the low-hanging fruit of using lists instead of loops and turning off checksums in existence checks.

Does this give enough speed-up for you?

**Comment by loveshack** [comment]
For what it's worth, I tried, but I can't get ara to work according to the instructions.  It always gives an error like this, independent of ARA_LOG_FILE.

```
 [WARNING]: Skipping plugin (/users/***/elasticluster/VENV/lib/python2.7
/site-packages/ara/plugins/callbacks/log_ara.py) as it seems to be invalid:
Unable to configure handler 'ara_file': [Errno 2] No such file or directory:
'/users/***/elasticluster/users/***/elasticluster/ara.log'
```
I haven't tried those commits, which I've only just seen, obviously.

**Comment by loveshack** [comment]
The current version is similar to what I got at 10-and-a-bit minutes to
bring up two c5 compute nodes using an AWS EL7 HPC image, thanks.  (This
is with the lmod role commented out.)  The time to "Configuring the
cluster" is ~2m20.  I also saved a minute or so previously when I tried
removing various playbooks for things I didn't want.

**Comment by riccardomurri** [comment]
You can also reduce setup time by forbidding upgrades and skipping installation of NIS/YP (if you're fine with running as the default OS user):

```
[setup/gridengine]
# do not upgrade installed software
global_var_upgrade_packages=no

# turn off NIS/YP
global_var_multiuser_cluster=no

# everything else stays the same...
```

(I guess you probably already figured this out, but I'm stating it anyway if someone else comes here via a web search.)

**Comment by riccardomurri** [comment]
> I also saved a minute or so previously when I tried removing various playbooks for things I didn't want.

What did you shave off? Since many people seem to be unhappy with the default set of installed software, I'm interesting in gathering data to make an option of software that's not deemed widely useful.

**Comment by riccardomurri** [comment]
> I tried, but I can't get ara to work according to the instructions. It always gives an error like this, independent of ARA_LOG_FILE.

Moved this thread to #618 (and fixed now).

**Comment by loveshack** [comment]
I should have said I can't reproduce the improvement I reported from commenting out irrelevant playbooks.  Presumably I didn't have good enough statistics when it looked significant, and I now see bigger variance in runtimes.

**Comment by yarikoptic** [comment]
Fwiw, I use [eatmydata](https://packages.debian.org/sid/utils/eatmydata)  to speed up package installation in containers. I think it could be of generic benefit here as well. Would be interested to hear on stats how much it helps eg on aws ;-)

**Comment by loveshack** [comment]
> Fwiw, I use
> [eatmydata](https://packages.debian.org/sid/utils/eatmydata) to speed
> up package installation in containers. I think it could be of generic
> benefit here as well. Would be interested to hear on stats how much it
> helps eg on aws ;-)

Good point, though I think the slowness is mainly down to ansible.  The
speed is quite variable anyway, presumably due to the network.  However,
I updated the libeatmydata rpms at
https://copr.fedorainfracloud.org/coprs/loveshack/livhpc/ in case that's
useful to anyone with a Fedora-ish system.

**Comment by riccardomurri** [comment]
I have done some tests with and without `eatmydata`, basically forcing
Ansible to use `/usr/bin/eatmydata python` instead of regular
`/usr/bin/python`.

Here are the results (MM:SS):

| setup | without `eatmydata` | *with* `eatmydata` |
| ----- | :------------------ | :----------------- |
| A     | 6:56.71             | 5:56.42            |
| B     | 6:56.66             | 5:58.53            |
| C     | 56:39.71            | ---                |

where the test setups are:

A. GridEngine, 1 master + 4 worker nodes, Debian 9.8 (official cloud image of 2019-03-21)
B. base OS + R 3.5, 1 server node, Debian 9.8 (official cloud image of 2019-03-21)
C. Spark/Hadoop, 1 master + 4 worker nodes, Ubuntu 16.04 (official cloud image of 2019-03-21)

In the case of C., I have interrupted the tests when I noticed that
some tasks had this error line:

       ERROR: ld.so: object 'libeatmydata.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.

I guess this is a bug in Ubuntu's packaging of `eatmydata`?

I will re-run the C. tests with Debian 9 next week.

**Comment by riccardomurri** [comment]
So it looks like `eatmydata` can give a good speedup in case of lengthy deployments with lots of packages (Spark, Jupyter) -- I'll package that up in a separate PR.

Thanks @yarikoptic for the suggestion!

**Comment by yarikoptic** [comment]
> I guess this is a bug in Ubuntu's packaging of `eatmydata`?

hm, unlikely packaging issue but probably some particularity of some execution, which may be first adjusts LD_LIBRARY_PATH or executes something in some kind of chroot while propagating environment variables? 
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=906927 confirms my suspicion for a probable cause

**Comment by riccardomurri** [comment]
`eatmydata` is now used by default starting version 1.3.dev13 (tagged yesterday); can be turned off with setup option `slow_but_safer`.

Thanks @yarikoptic for the suggestion!

**Comment by riccardomurri** [comment]
It may be relevant to this thread that ElastiCluster ("master" branch) can use the [mitogen][1] library if it is installed.  Using mitogen speeds up creation of a GridEngine on Debian cluster from ~6 minutes to ~4 minutes 30 seconds; I haven't tried measured other cluster types.

Most of the work upon initial setup is installing software packages, which won't be sped up much by mitogen (or any other approach, for the matter) but resizes or reconfigurations, when most of the Ansible tasks are a no-op, should take much more advantage of Mitogen.

[1]: https://mitogen.networkgenomics.com/ansible_detailed.html
